knodle/transformation/majority.py

```diff
@@ -34,6 +35,30 @@ def probabilies_to_majority_vote(
         raise ValueError("Specify a way how to resolve unclear majority votes.")
 
 
+def z_matrix_to_rule_idx(
+        rules: np.ndarray, choose_random_rule: bool = True
+) -> int:
+    """Transforms a z matrix to rule indices of matching rules.
+    If there is more than one rule match: either select one of the rules randomly or return a vector containing
+    all of them.
+
+    Args:
+        rules: Vector of probabilities for 1 sample. Shape: classes x 1
+        choose_random_rule: Choose a random label, if there's no clear majority.
+    Returns: An array of classes.
+    """
+
+    row_max = np.max(rules)
+    num_occurrences = (row_max == rules).sum()
+    if num_occurrences == 1:
+        return int(np.argmax(rules))
+    elif choose_random_rule:
+        max_ids = np.where(rules == row_max)[0]
+        return int(np.random.choice(max_ids))
+    else:
+        raise ValueError("Specify a way how to resolve multiple rule matches.")
```

Comment:
not sure about `else`. Why not simply: 
```
if num_occurrences == 1:
    return int(np.argmax(rules))
else:
    max_ids = np.where(rules == row_max)[0]
    return int(np.random.choice(max_ids))
```
I like the idea of giving an option how more-than-one-rule-match cases should be handled, but since the only one option we provide is random, doesn't`ValueError` look strange? Hm.

Reply:
Have an idea: how about providing an option to filter such samples out? currently, we don't have this use case in our experiments, but I can imagine someone would need it. And also later there will be also an option to add the duplicated samples with one rule matched in each sample, right? mb then:
```
def z_matrix_to_rule_idx(rules: np.ndarray, filter_multiple_matches: bool = True, choose_random_rule: bool = True, dublicate_samples: bool = False
...
     if num_occurrences == 1:
         return int(np.argmax(rules))
     else:
         if filter_multiple_matches:
            ...
         elif choose_random_rule:
            ...
         else:
            raise ValueError("Multiple rule matches can be either filtered out or duplicated. Please specify.")
```

Reply:
I implemented it also w.r.t to the later duplicate_instances option that would give an instance with every rule match. But I am not sure, whether this would be a good idea (could result in a giant dataset).

Reply:
that's true! But I think, as a framework, we could still provide this functionality (but don't use it by default), and let the user to decide whether to use it or not. 

We could come back to this discussion tomorrow